src/libos/crates/async-rt/src/wait/macros.rs

```diff
@@ -6,21 +6,39 @@ use super::{Waiter, WaiterQueue};
 /// is used to avoid busy loop. The user code of the loop body can use
 /// `break`, `continue`, or `return` as it would do in a normal loop.
 ///
+/// However, there are restrictions on the return value, the user code
+/// must return a `Result` when it `break` or `return`. If the user code
+/// doesn't use `break` and the waiter_loop isn't the return value, the user
+/// code need to use explicit annotation to tell compiler the type of `Result`.
+/// ```
+///
+/// ```
+///
 /// Between each rounds of loop (e.g., via `continue`), the execution will
 /// pause until the waiter queue given to the macro call wakes up the waiter
 /// associated with this loop.
 ///
 /// This macro is more preferable than using Waiter and WaiterQueue directly.
 /// Not only because the macro is more easy to use, but also because it is
 /// much harder to be misuse.
+///
+/// The waiter_loop also support timeout. if user privide a timeout, this loop
+/// will be forced to exit when the timeout expired. When the the timeout expired,
+/// the loop will break, user should deal with the timeout case by themselves.
+/// The timeout argument should be an Option of Duration, e.g., `Option<Duration>`
+/// `Option<&mut Duration>`.
 #[macro_export]
 macro_rules! waiter_loop {
     ($waiter_queue:expr, $loop_body:block) => {{
-        use $crate::wait::{AutoWaiter, Waiter, WaiterQueue};
+        $crate::waiter_loop!($waiter_queue, None::<core::time::Duration>, $loop_body)
+    }};
+    ($waiter_queue:expr, $timeout:expr, $loop_body:block) => {{
+        use $crate::wait::{AutoWaiter, WaiterQueue};
 
         let waiter_queue: &WaiterQueue = $waiter_queue;
         let mut auto_waiter = AutoWaiter::new(waiter_queue);
         let mut loop_count: u32 = 0;
+        let mut timeout = $timeout.map_or(None, |t| Some(t));
```

Comment:
What is the purpose of `map_or` here? The `map_or` does not seem to have any effect.

Reply:
the purpose is to create a `mut timeout` if the timeout is not `mut`